snowdrop-execution/src/Snowdrop/Execution/DbActions/Simple.hs

```diff
@@ -17,154 +16,201 @@ import           Universum
-        ChangeSet <$> foldM processOne mempty (M.toList cs)
+    injUndo :: HChangeSet xs -> undo
+    injUndo = inj
 
-sumChangeSetDaa
-    :: (IdSumPrefixed id, Ord id, Applicative m)
-    => (StateR id -> m (StateP id value))
-    -> (forall b . Prefix -> b -> ((id, value) -> b -> b) -> m b)
-    -> DbAccessActions (SumChangeSet id value) id value m
-sumChangeSetDaa getterImpl iterImpl =
-    DbAccessActions (chgAccumGetter getterImpl) (chgAccumIter iterImpl)
+    computeUndoDo
+        :: SumChangeSet xs
+        -> HChangeSet xs
+        -> m (Either CSMappendException (HChangeSet xs))
+    computeUndoDo = computeHChSetUndo (daaGetter daa)
 
 chgAccumIter
-    :: (IdSumPrefixed id, Ord id, Applicative m)
-    => (Prefix -> b -> ((id, value) -> b -> b) -> m b)
-    -> SumChangeSet id value
-    -> Prefix
-    -> b
-    -> ((id, value) -> b -> b)
-    -> m b
-chgAccumIter iter accum prefix initB foldF =
-    let newKeysB = M.foldrWithKey (\i v r -> foldF (i, v) r) initB $ getNewKeys accum prefix
-        newFoldF (i, val) b = case queryAccumOne accum i of
-            Nothing         -> foldF (i, val) b
-            Just Rem        -> b
-            Just NotExisted -> b -- something strange happened -- TODO shall we throw error here ?
-            Just (Upd newV) -> foldF (i, newV) b
-            Just (New _)    -> b -- something strange happened
-     in iter prefix newKeysB newFoldF
+    :: (Applicative m, RecAll' xs OrdHKey)
+    => DIter' xs m
+    -> DIter (SumChangeSet xs) xs m
+chgAccumIter RNil (SumChangeSet RNil) = pure RNil
+chgAccumIter (iter' :& xs) (unSumCS -> csel' :& ys) =
+    (chgAccumIter' iter' csel' :&) <$> chgAccumIter xs (SumChangeSet ys)
+  where
+    chgAccumIter' :: OrdHKey t => IterAction m t -> HChangeSetEl t -> IterAction m t
+    chgAccumIter' iter ac'@(HChangeSetEl accum) = IterAction $ \initB foldF -> do
+        let newKeysB = M.foldrWithKey (\i v r -> foldF (i, v) r) initB $ csNew ac'
+            newFoldF (i, val) b = case M.lookup i accum of
+                Nothing         -> foldF (i, val) b
+                Just Rem        -> b
+                Just NotExisted -> b -- TODO shall we throw error here ?
+                Just (Upd newV) -> foldF (i, newV) b
+                Just (New _)    -> b -- TODO shall we throw error here?
+        runIterAction iter newKeysB newFoldF
+chgAccumIter _ _ = error "Absurd. Lol, without this I get non exhaustive pattern matching "
 
 chgAccumGetter
-    :: (IdSumPrefixed id, Ord id, Applicative m)
-    => (StateR id -> m (StateP id value))
-    -> SumChangeSet id value
-    -> StateR id
-    -> m (StateP id value)
+    :: (Applicative m, HIntersectable xs xs, Semigroup (HMap xs))
+    => DGetter' xs m
+    -> DGetter (SumChangeSet xs) xs m
 chgAccumGetter getter accum reqIds =
-    bool (unionStateP resp <$> getter reqIds') (pure resp) (null reqIds')
+    bool (unionStateP resp <$> getter reqIds') (pure resp) (rnull reqIds')
```

Comment:
Are you sure `rnull` is what you want here?
My impresion is that you will always get `rnull (hintersect x y)` equal to `False` as instead of checking every component contains empty map you try to check that there is no components.